volume/store/store.go

```diff
@@ -47,39 +60,94 @@ func (s *VolumeStore) AddAll(vols []volume.Volume) {
 func (s *VolumeStore) Create(name, driverName string, opts map[string]string) (volume.Volume, error) {
 	s.mu.Lock()
 	if vc, exists := s.vols[name]; exists {
-		v := vc.Volume
 		s.mu.Unlock()
-		return v, nil
+		<-vc.waitCreate
+		if vc.Volume != nil && (vc.Volume.DriverName() == driverName || driverName == volume.DefaultDriverName) {
+			return vc.Volume, nil
+		}
+		return nil, ErrNameConflict
 	}
+
+	vc := newVolumeCounter()
+	s.vols[name] = vc
 	s.mu.Unlock()
 	logrus.Debugf("Registering new volume reference: driver %s, name %s", driverName, name)
 
-	vd, err := volumedrivers.GetDriver(driverName)
+	var (
+		err error
+		vd  volume.Driver
+		v   volume.Volume
+	)
+
+	defer func() {
+		if err != nil {
+			s.mu.Lock()
+			delete(s.vols, name)
+			s.mu.Unlock()
+		}
+		close(vc.waitCreate)
```

Comment:
This may result in deadlock.
1. plugin1 is creating volume "test"
2. plugin2 try to create "test", "test" exist; so plugin2 hold the lock and wait for vc.waitCreate
3. plugin1 create "test" failed, and come to delete(s.vols, name) trying to get muLock, which is held by plugin2

move this close() forward may solve this problem. 
Anyway, plugin2 create() won't work regardless plugin1 create() succeed or not.